awss3/store.go

```diff
@@ -0,0 +1,703 @@
+
+	itemLimit := int64(f.PageSize)
+	if q.PageSize > 0 {
+		itemLimit = int64(q.PageSize)
+	}
+
+	params := &s3.ListObjectsInput{
+		Bucket:  aws.String(f.bucket),
+		Marker:  &q.Marker,
+		MaxKeys: &itemLimit,
+		Prefix:  &q.Prefix,
+	}
+
+	resp, err := f.client.ListObjects(params)
+	if err != nil {
+		u.Warnf("err = %v", err)
+		return nil, err
+	}
+
+	objResp := &cloudstorage.ObjectsResponse{
+		Objects: make(cloudstorage.Objects, len(resp.Contents)),
+	}
+
+	for i, o := range resp.Contents {
+		objResp.Objects[i] = newObject(f, o)
+	}
+
+	if *resp.IsTruncated {
+		q.Marker = *resp.Contents[len(resp.Contents)-1].Key
+	}
+
+	return objResp, nil
+}
+
+// Objects returns an iterator over the objects in the google bucket that match the Query q.
+// If q is nil, no filtering is done.
+func (f *FS) Objects(ctx context.Context, q cloudstorage.Query) (cloudstorage.ObjectIterator, error) {
+	return cloudstorage.NewObjectPageIterator(ctx, f, q), nil
+}
+
+// Folders get folders list.
+func (f *FS) Folders(ctx context.Context, q cloudstorage.Query) ([]string, error) {
+
+	q.Delimiter = "/"
+
+	// Think we should just put 1 here right?
+	itemLimit := int64(f.PageSize)
+	if q.PageSize > 0 {
+		itemLimit = int64(q.PageSize)
+	}
+
+	params := &s3.ListObjectsInput{
+		Bucket:    aws.String(f.bucket),
+		MaxKeys:   &itemLimit,
+		Prefix:    &q.Prefix,
+		Delimiter: &q.Delimiter,
+	}
+
+	folders := make([]string, 0)
+
+	for {
+		select {
+		case <-ctx.Done():
+			// If has been closed
+			return folders, ctx.Err()
+		default:
+			if q.Marker != "" {
+				params.Marker = &q.Marker
+			}
+			resp, err := f.client.ListObjectsWithContext(ctx, params)
+			if err != nil {
+				return nil, err
+			}
+			for _, cp := range resp.CommonPrefixes {
+				folders = append(folders, strings.Trim(*cp.Prefix, `/`))
+			}
+			return folders, nil
+		}
+	}
+}
+
+/*
+// Copy from src to destination
+func (f *FS) Copy(ctx context.Context, src, des cloudstorage.Object) error {
+
+	so, ok := src.(*object)
+	if !ok {
+		return fmt.Errorf("Copy source file expected s3 but got %T", src)
+	}
+	do, ok := des.(*object)
+	if !ok {
+		return fmt.Errorf("Copy destination expected s3 but got %T", des)
+	}
+
+	oh := so.b.Object(so.name)
+	dh := do.b.Object(do.name)
+
+	_, err := dh.CopierFrom(oh).Run(ctx)
+	return err
+}
+
+// Move which is a Copy & Delete
+func (f *FS) Move(ctx context.Context, src, des cloudstorage.Object) error {
+
+	so, ok := src.(*object)
+	if !ok {
+		return fmt.Errorf("Move source file expected s3 but got %T", src)
+	}
+	do, ok := des.(*object)
+	if !ok {
+		return fmt.Errorf("Move destination expected s3 but got %T", des)
+	}
+
+	oh := so.b.Object(so.name)
+	dh := do.b.Object(des.name)
+
+	if _, err := dh.CopierFrom(oh).Run(ctx); err != nil {
+		return err
+	}
+
+	return oh.Delete(ctx)
+}
+*/
+// NewReader create GCS file reader.
+func (f *FS) NewReader(o string) (io.ReadCloser, error) {
```

Comment:
I would like to deprecate these non-ctx methods and remove them in the future.

Reply:
sounds like all you!